lightning/src/offers/static_invoice.rs

```diff
@@ -0,0 +1,1130 @@
+			created_at: Some(self.created_at.as_secs()),
+			relative_expiry: self.relative_expiry.map(|duration| duration.as_secs() as u32),
+			fallbacks: self.fallbacks.as_ref(),
+			features,
+			node_id: Some(&self.signing_pubkey),
+		}
+	}
+
+	fn chain(&self) -> ChainHash {
+		debug_assert_eq!(self.offer.chains().len(), 1);
+		self.offer.chains().first().cloned().unwrap_or_else(|| self.offer.implied_chain())
+	}
+
+	fn metadata(&self) -> Option<&Vec<u8>> {
+		self.offer.metadata()
+	}
+
+	fn amount(&self) -> Option<Amount> {
+		self.offer.amount()
+	}
+
+	fn offer_features(&self) -> &OfferFeatures {
+		self.offer.features()
+	}
+
+	fn description(&self) -> Option<PrintableString> {
+		self.offer.description()
+	}
+
+	fn absolute_expiry(&self) -> Option<Duration> {
+		self.offer.absolute_expiry()
+	}
+
+	fn issuer(&self) -> Option<PrintableString> {
+		self.offer.issuer()
+	}
+
+	fn offer_message_paths(&self) -> &[BlindedPath] {
+		self.offer.paths()
+	}
+
+	fn message_paths(&self) -> &[BlindedPath] {
+		&self.message_paths[..]
+	}
+
+	fn supported_quantity(&self) -> Quantity {
+		self.offer.supported_quantity()
+	}
+
+	fn payment_paths(&self) -> &[(BlindedPayInfo, BlindedPath)] {
+		&self.payment_paths[..]
+	}
+
+	fn created_at(&self) -> Duration {
+		self.created_at
+	}
+
+	fn relative_expiry(&self) -> Duration {
+		self.relative_expiry.unwrap_or(DEFAULT_RELATIVE_EXPIRY)
+	}
+
+	#[cfg(feature = "std")]
+	fn is_expired(&self) -> bool {
+		is_expired(self.created_at(), self.relative_expiry())
+	}
+
+	fn fallbacks(&self) -> Vec<Address> {
+		let chain =
+			self.offer.chains().first().cloned().unwrap_or_else(|| self.offer.implied_chain());
+		self.fallbacks
+			.as_ref()
+			.map(|fallbacks| filter_fallbacks(chain, fallbacks))
+			.unwrap_or_else(Vec::new)
+	}
+
+	fn features(&self) -> &Bolt12InvoiceFeatures {
+		&self.features
+	}
+
+	fn signing_pubkey(&self) -> PublicKey {
+		self.signing_pubkey
+	}
+}
+
+impl Writeable for StaticInvoice {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), io::Error> {
+		WithoutLength(&self.bytes).write(writer)
+	}
+}
+
+impl TryFrom<Vec<u8>> for StaticInvoice {
+	type Error = Bolt12ParseError;
+
+	fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+		let parsed_invoice = ParsedMessage::<FullInvoiceTlvStream>::try_from(bytes)?;
+		StaticInvoice::try_from(parsed_invoice)
+	}
+}
+
+tlv_stream!(InvoiceTlvStream, InvoiceTlvStreamRef, 160..240, {
+	(160, payment_paths: (Vec<BlindedPath>, WithoutLength, Iterable<'a, BlindedPathIter<'a>, BlindedPath>)),
+	(162, blindedpay: (Vec<BlindedPayInfo>, WithoutLength, Iterable<'a, BlindedPayInfoIter<'a>, BlindedPayInfo>)),
+	(164, created_at: (u64, HighZeroBytesDroppedBigSize)),
+	(166, relative_expiry: (u32, HighZeroBytesDroppedBigSize)),
+	(172, fallbacks: (Vec<FallbackAddress>, WithoutLength)),
+	(174, features: (Bolt12InvoiceFeatures, WithoutLength)),
+	(176, node_id: PublicKey),
+	(178, message_paths: (Vec<BlindedPath>, WithoutLength)),
+});
```

Comment:
Should we re-use the TLV stream from `invoice.rs`? That was the approach that we used for `Refund` even though the fields have different requirements.

Reply:
Hmm, my reasoning was that this is <10 lines and leans on the type system whereas manual checks for `payment_hash`, amount, and message paths would be more verbose/error prone and require new `Bolt12SemanticError` variant(s). Do you think it's better to reuse it?